data/tensorflow/operator/recordio_dataset_op.cc

```diff
@@ -0,0 +1,215 @@
+#include "tensorflow/core/framework/dataset.h"
+#include "tensorflow/core/framework/common_shape_fns.h"
+#include "../recordio/recordio_reader.h"
+
+using namespace tensorflow;
+
+REGISTER_OP("RecordioDataset")
+    .Input("filenames: string")
+    .Input("start_chunk: int64")
+    .Input("chunk_count: int64")
+    .Output("handle: variant")
+    .SetIsStateful()  
+    .SetShapeFn([](shape_inference::InferenceContext* c) {
+      shape_inference::ShapeHandle unused;
+      // `filenames` must be a scalar or a vector.
+      TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(0), 1, &unused));
+      // `start_chunk` could only be a scalar.
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 0, &unused));
+      // `chunk_count` could only be a scalar.
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 0, &unused));
+      return shape_inference::ScalarShape(c);
+    });
+
+class RecordIODatasetOp : public DatasetOpKernel {
+ public:
+  using DatasetOpKernel::DatasetOpKernel;
+
+  void MakeDataset(OpKernelContext* ctx, DatasetBase** output) override {
+    const Tensor* filenames_tensor;
+    OP_REQUIRES_OK(ctx, ctx->input("filenames", &filenames_tensor));
+    OP_REQUIRES(
+        ctx, filenames_tensor->dims() <= 1,
+        errors::InvalidArgument("`filenames` must be a scalar or a vector."));
+
+    std::vector<string> filenames;
+    filenames.reserve(filenames_tensor->NumElements());
+    for (int i = 0; i < filenames_tensor->NumElements(); ++i) {
+      filenames.push_back(filenames_tensor->flat<string>()(i));
+    }
+
+    int64 start_chunk = -1;
+    OP_REQUIRES_OK(
+        ctx, ParseScalarArgument<int64>(ctx, "start_chunk", &start_chunk));
+    OP_REQUIRES(ctx, start_chunk>= 0,
+                errors::InvalidArgument(
+                    "`start_chunk` must be >= 0"));
+
+    int64 chunk_count = -1;
+    OP_REQUIRES_OK(
+        ctx, ParseScalarArgument<int64>(ctx, "chunk_count", &chunk_count));
+    OP_REQUIRES(ctx, chunk_count>= 0,
+                errors::InvalidArgument(
+                    "`chunk_count` must be >= 0 (0 means read all the chunks)"));
+
+    *output =
+        new Dataset(ctx, std::move(filenames), start_chunk, chunk_count);
+  }
+
+ private:
+  class Dataset : public GraphDatasetBase {
+   public:
+    explicit Dataset(OpKernelContext* ctx, std::vector<string> filenames,
```

Comment:
pass filenames use const std::vector<string>&

Reply:
already change to const string& filename